fix the wrong msg_col after calling execute (issue 6035/9250)

liushapku opened the pull request:
addresses [6035](https://github.com/neovim/neovim/issues/6035) and [9250](https://github.com/neovim/neovim/issues/9250)

liushapku committed:
UI: Fix wrong msg_col after execute()

closes #6035
closes #9250

bfredl commented on src/nvim/eval.c:
```diff
@@ -8132,6 +8132,8 @@ static void f_execute(typval_T *argvars, typval_T *rettv, FunPtr fptr)
   }
 
+  int save_msg_col = msg_col;
+  int should_restore_col = 0;
```
use `bool` `true` and `false`

liushapku replied:
Thanks for pointing out that. I almost forget c has bool. 

Since the PR has been created, what should I do now? 

1. Cancel and create a new one?
2. Rely on the one to merge it to update it?

bfredl replied:
just update the PR. force-pushing is okay.

bfredl commented:
Looks reasonable, a test in `test/functional/eval/execute_spec.lua` would be good. Visual test is probably the best way to check the behavior, there are already some in the test file.

justinmk commented:
Thanks @liushapku . Do you plan to add a test?

cc @jamessan who worked in this area IIRC

liushapku commented:
> Thanks @liushapku . Do you plan to add a test?

@justinmk Sorry I don't know lua yet. So I am not able to do that. 

By the way, when someone is to implement the test, he/she may also consider adding another 4 test with the `echon` in the third line of the functions replaced by `echo`.

justinmk commented:
@liushapku  `Test1a` and `Test1b` have the same result before/after your change, is that expected?

I pushed an update to this PR. Do the tests look correct?

liushapku commented:
@justinmk, `Test1b` is the benchmark and `Test1a`'s behavior is not changed. Since `should_restore_col == false` when the second paramter of `execute(command, mode)` is `mode = ''`. In this case, we should respect the output and status change from within the execution of `command`. 

See also a recent comment I made in [`vim`](https://github.com/vim/vim/commit/10ccaa17ec8b2be1132fd19059e1cd5fb5c902c4). There I also mentioned another [bug](https://github.com/vim/vim/issues/3659) of the behavior of `execute()`, which is also shared by `neovim`. I think I got a solution to it and is working on it. But I didn't duplicate the issue here under `neovim`. 

By the way, may I ask for the general procedure of raise/handling an issue that is shared by both `vim` and `neovim`? Should I:

1. raise at both repos or 
2. raise at `vim` site first and wait it to be solved to create a `vim-patch` request in `neovim`;

justinmk commented:
@lincheney In this case, don't force-push. Because the current state was already reviewed. I will squash it when merging.

https://github.com/vim/vim/issues/3659 might be annoying to implement (could require another global) for not much benefit.

> general procedure of raise/handling an issue that is shared by both vim and neovim

Ideally, we want Vim bugs to be fixed by Vim. Then it's greatly appreciated if you [send us the patch](https://github.com/neovim/neovim/wiki/Merging-patches-from-upstream-Vim) after that.

liushapku commented:
I have got a solution not to introduce new global variable but to reinterpret `redir_off` in a different way. I will update you with the status.

liushapku commented:
> for not much benefit.
> 
> > general procedure of raise/handling an issue that is shared

liushapku commented:
Accidentally clicked `Close and comment`. Sorry for the noise created.

liushapku commented:
> @justinmk Thanks for the revision to the PR. I would like slightly modify the test cases. Would you please teach me how to make a force push to a **PR** and how to pull your changes to the **PR**?

@justinmk I fetched the repo and didn't get the commit of your forced push to the PR. What should I to download/sync that commit locally and then work on it? Should I save the patch and manually apply patch or there is more straight forward way to do it. I want make the changes to the PR. I also need to learn how to force push my updates back. Thanks.

justinmk commented:
If you `git fetch --all` then the commit 2ff49a42ba3bac477afdeb0f0f64c568b9effe03 will exist _somewhere_ in your local repo, so you can always get it directly.

```
git fetch --all
git checkout -b pr9319
git reset --hard 2ff49a42ba3bac477afdeb0f0f64c568b9effe03
```

> What should I to download/sync that commit locally

`git fetch --all` always gets _everything_ from the remote into your local database. So at the very least you can always refer to the literal commit hashes.

> I also need to learn how to force push my updates back

Since you pushed to `master` for this PR, you need to push to `master` to update the PR.

```
git push --force origin HEAD:master
```

justinmk commented:
maybe `did_emsg` .

liushapku committed:
update functional test for "places cursor correctly #6035"

liushapku commented:
I updated code to handle the discussed bug and also the tests. 

For your review: the major changes are

- removed all `Testxb` tests since they do not involve `execute()`
- added a few cases (read the comment you will know what is being tested)
- `echo` inside `execute()` is changed to `echon` since if `echon` is more sensitive to the change of `msg_col`

Of particular interest is `Test4` and `Test7`.

- `Test4`: `echoerr` uses `f_execute` as its `.cmd_func`, which does not set `did_emsg`, **This is a corner case for which I still didn't get a way to handle it correctly.** Your input is much appreciated.
- `Test7`: after an existing error so that `did_emsg` is already set before calling `execute()`, `echon` starts from a newline. This maybe intentional and may be a bug from `vim`. Because the *correct behavior* is not clearly defined after error. **So I feel this case needs no further treatment; Do you also agree?**

liushapku commented:
> [vim/vim#3659](https://github.com/vim/vim/issues/3659) might be annoying to implement (could require another global) for **not much benefit**.

@justinmk, I have made a more realistic [example](https://github.com/vim/vim/issues/3659#issuecomment-445460921) to justify my concern for issue #3659

justinmk commented on src/nvim/eval.c:
```diff
@@ -8141,19 +8144,23 @@ static void f_execute(typval_T *argvars, typval_T *rettv, FunPtr fptr)
   ga_init(&capture_local, (int)sizeof(char), 80);
   capture_ga = &capture_local;
   redir_off = false;
+  const bool save_did_emsg = did_emsg;
```
It's customary to keep the `save_xx` declarations grouped together (line 8126) if possible.

justinmk commented:
One of Vim's tests (which we also run, besides our Lua tests) is failing:

```
Found errors in Test_execute_string():
function RunTheTest[35]..Test_execute_string line 3: Expected 'noendofline' but got '                           noendofline'
```

You can run that test locally like this:

    NVIM_LISTEN_ADDRESS= VIMRUNTIME= TEST_FILE=test_execute_func.res make oldtest

liushapku commented:
@justinmk OK. I will follow what you suggested. Lets move on to finish the case.

Now the only problem that hinders the progress is that the PR breaks the old test. I have tested in my computer and confirmed the follows:

1. The test fails if I `NVIM_LISTEN_ADDRESS= VIMRUNTIME= TEST_FILE=test_execute_func.res make oldtest`. The reason is

> <pre>function RunTheTest[35]..Test_execute_string line 3: Expected 'noendofline' but got '         noendofline'</pre>

2. However, if I run `build/bin/nvim -u` [src/nvim/testdir/test_execute_func.vim](https://github.com/neovim/neovim/blob/master/src/nvim/testdir/test_execute_func.vim) and type `:call Test_execute_string() | echo v:errors` it does not show error.
3. If I save the following three lines in a file and source the file, `x`, `y` and `z` are what they are compared against. (the failed test is because `z != "noendofline"`, but here `z == "noendofline"`)
```vim
let x = execute('set compatible?')
let y = execute('echo "something\nnice"')
let z = execute('echon "noendofline"')
```

*So I don't know what is missing in the test environment that has resulted in the different result.*

@janlazo  is my finding the same as [yours](https://github.com/neovim/neovim/issues/6035#issuecomment-445482969)?

janlazo commented:
> @janlazo is my finding the same as [yours](https://github.com/neovim/neovim/issues/6035#issuecomment-445482969)

@liushapku No. I pushed my port of 8.1.0569 and 8.1.0571 in https://github.com/neovim/neovim/issues/6035#issuecomment-445482969. Running `TEST_FILE=test_execute_func.res make oldtest`, I get the following log:

```
From test_execute_func.vim:
Found errors in Test_execute_does_not_change_col():
function RunTheTest[35]..Test_execute_does_not_change_col line 9: Expected 'abcdxyz' but got '       '
Found errors in Test_execute_not_silent():
function RunTheTest[35]..Test_execute_not_silent line 9: Expected 'abcd234 ' but got '        '
function RunTheTest[35]..Test_execute_not_silent line 14: Expected 'xyz ' but got '    '
```

Running it manually with `nvim -u NONE -S src/nvim/testdir/test_execute_func.vim -c 'call Test_execute_does_not_change_col()' -c 'call Test_execute_not_silent()'`, v:errors  has the folllowing:

```
['function Test_execute_does_not_change_col line 9: Expected ''abcdxyz'' but got
 ''xyz    ''']
```

justinmk commented:
> @justinmk Please make a decision whether to go on with this PR or adopt the vim patches

Better to use Vim's approach if there's no strong reason to do otherwise.

The tests added in this PR are still worth including, though they may need to be adjusted.

liushapku commented:
@justinmk Sure. Please do that then. We can finally close it.

justinmk commented:
This isn't high on my priority list, so if it's up to me to do it, it won't get done.

liushapku commented:
It is still up to you

justinmk commented:
Then forget it.

liushapku commented:
That would be the best. Thank you.

janlazo commented:
Opened https://github.com/neovim/neovim/pull/9338
